app/db/models.py

```diff
@@ -1,22 +1,68 @@
 """Database models for the application."""
 
+from typing import Any
 from uuid import uuid4
 
-from sqlalchemy import String
+from loguru import logger
+from sqlalchemy import String, event, text
 from sqlalchemy.dialects.postgresql import UUID
+from sqlalchemy.engine import Connection
+from sqlalchemy.exc import SQLAlchemyError
 from sqlalchemy.orm import Mapped, mapped_column
 
 from app.db.base import Base
 from app.db.model_mixins import TimestampMixin
 
+NOTIFICATION_STARTING_PARTITION_YEAR = 2024
+NOTIFICATION_PARTITION_DURATION_YEARS = 10
+
 
 class Notification(TimestampMixin, Base):
     """Database table for notifications."""
 
     __tablename__ = 'notifications'
-
-    id: Mapped[UUID[str]] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid4)
-    personalization: Mapped[str] = mapped_column(String, nullable=True)
+    __table_args__ = {'postgresql_partition_by': 'RANGE (created_at)'}  # noqa: RUF012
+    __mapper_args__ = {'primary_key': ['id', 'created_at']}  # noqa: RUF012
+
+    id: Mapped[UUID[str]] = mapped_column(UUID(as_uuid=True), nullable=False, default=uuid4)
+    personalization: Mapped[str | None] = mapped_column(String, nullable=True)
+
+
+@event.listens_for(
+    Notification.__table__,
+    'after_create',
+)
+def create_notification_year_partition(target: Base, connection: Connection, **kwg: dict[str, Any]) -> None:
```

Comment:
I just have a question about the use of `Any` here. Is this necessary for SQLAlchemy?

Reply:
Yes - It passes in meta data. If I take out `kwg`, the method errors out. It has to be ANY because the arg is actually coming from the third-party library SQL Alchemy and SQL Achemy defines `kwg` as Any.

https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.DDLEvents.after_create:~:text=Called%20after%20CREATE%20statements%20are%20emitted.